### PR TITLE
Map Rendering with Shaders

### PR DESCRIPTION
**Pull request by speveril**
The whole thing just renders into a &lt;canvas>. It should work with whatever styling applied to it; you might need to call .resize on the map if you've done something to resize it (it will automatically handle window resizes, but if it gets placed inside something else that may not work). The canvas is the size of the viewport; the map scrolls within that viewport. Because of that it may be hard to get scroll bars on the canvas.

That said, it currently supports moving the map around (click-drag) and zooming (scroll-wheel). app.js instantiates a map.

I haven't changed the map JSON formats at all. It seems like it'd make sense to have the reference to the map/tile data and to the vsp data from inside the original map JSON. That shouldn't be too hard to change.

### CODE REVIEW

File: app/map.js [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    gl.attachShader(program, fragmentShader);
+    gl.linkProgram(program);
+    if (!gl.getProgramParameter(program, gl.LINK_STATUS)) {
+        throw new Error("Failed to link shader program: " + gl.getProgramInfoLog(program));
+    }
+
+    return program;
+}
+
+function buildTileDataTexture(data) {
```
Comment (mcgrue) [inline app/map.js]: @speveril can you explain this function?  Bitshifting and moduloing sounds like modscene efficiency and I'm too sick to think atm.

(And too lazy to think at other times.)

File: app/map.js [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+    this.mapPath = mapfile;
+    this.mapData = jetpack.read(mapfile, 'json');
+    this.renderString = this.mapData.renderstring.split(",");
+    this.mapSize = [0,0];
+    for (var i = 0; i < this.mapData.layers.length; i++) {
+        if (this.mapData.layers[i].dimensions.X > this.mapSize[0]) this.mapSize[0] = this.mapData.layers[i].dimensions.X;
+        if (this.mapData.layers[i].dimensions.Y > this.mapSize[1]) this.mapSize[1] = this.mapData.layers[i].dimensions.Y;
+    }
+    this.camera = [0, 0, 1];
```
Comment (mcgrue) [inline app/map.js]: The camera is in 3 dimensional space but we're just treating it like it's 2d?

what do the ints represent? "units"?

File: app/map.js (hunk 1 above)
Comment (speveril) [inline app/map.js]: It's building a texture where each pixel actually represents a tile index. The texture, though, is colors, defined as 4 x 8bit ints. If we want to support more than 256 possible tiles, we've got to break the actual tile index apart into its constituent bytes, and then store then as consecutive Uint8s in the array. 

(t >> 8) % 256 is just "get the byte starting at bit 8".

File: app/map.js (hunk 2 above)
Comment (speveril) [inline app/map.js]: Not quite -- this.camera is [x, y, scale factor]. The x and y coordinates are pixel-based, and the scale factor is just a float from 0.125 to 16 (in the current zoom handling anyway).

File: app/map.js (hunk 1 above)
Comment (mcgrue) [inline app/map.js]: I saw this technique in other tutorials.  It somewhat baffles me that it's how things are really done.

I suppose textures are the data format du jour for shaders?

**Comment by mcgrue** [comment]
LGTM

File: app/map.js (hunk 1 above)
Comment (speveril) [inline app/map.js]: When I was working on it I had a thought that you might be able to do it with an attribute array, the same way that the position data is passed in. However, textures seem to be the best way to handle very large arrays of homogeneous data like this, so I just stuck with the texture for now.

**Pull request merged**